test/test_distributions.py

```diff
@@ -1651,9 +1652,10 @@ def inv_vec_transform(y):
             ),
             (2,),
         ),
+        (Simplex2OrderedTransform(0.0),(5,))
     ],
 )
-@pytest.mark.parametrize("batch_shape", [(), (1,), (3,), (6,), (3, 1), (1, 3), (5, 3)])
+@pytest.mark.parametrize("batch_shape", [pytest.param((), marks=pytest.mark.xfail(reason="Simplex2Ordered cannot run with batch_shape=()")), (1,), (3,), (6,), (3, 1), (1, 3), (5, 3)])
```

Comment:
To resolve this issue, I guess you can test for
```
ComposedTransform([biject_to(constraints.simplex), Simplex2OrderedTransform(0.0), biject_to(constraints.ordered_vector).inv])
```
This one will test for the transform: `x_unconstrained -> (x -> y ->) y_unconstrained` instead of `x -> y`.

Reply:
Would you mind saying a bit more about this proposal?

It seems similar to what `test_bijective_transforms` does, ie, Normal--> biject to domain (=x) --> SimplexToOrdered (=y) and then we test.

I've tried to plug it into parameterized but it also fails in the jacobian step - this time not on the square shape, but on the value itself (`actual J != excepted J`)

Reply:
Sorry, I should be clearer. I meant to replace
```python
(Simplex2OrderedTransform(0.0),(5,))
```
by
```python
(ComposedTransform([biject_to(constraints.simplex), Simplex2OrderedTransform(0.0), biject_to(constraints.ordered_vector).inv]), (5,))
```
The test will compute log det jacobian of the composed transform implementation, and compare it with the autodiff one.

Reply:
> actual J != excepted J

Your implementation looks like the one in Stan's implementation. I'm not sure what's going wrong. The derivation of jacobian seems not trivial to me. I'll look into details later in the week. One thing I'm not sure in the Stan implementation is why Jacobian has shape 5 x 5. Shouldn't it have shape 4 x 4?

Reply:
Yes, that's exactly what I've done and I was confused by the error as well.

As for the Jacobian - that puzzled me too at first! 
It's a square matrix because the map is `p` of dim 5 --> (`c` of dim 4, S ), where S is a variable representing the normalization (sum of p = 1); That's why the left column in Jacobian is ones -- because dp\dS is 1
Together it's a 5x5 map.

> Case Study 2.2: ... This allows us to transform the K probabilities (p1,…,pK) to a new variable that encodes the normalization, S=∑Kk=1pk=1 and the K−1 cut points, (S,c1,…,cK−1).... 

It doesn't matter for the density afterwards, because it's marginalized out

*** 
Not sure what to do with the autodiff - I have verified the Jacobian earlier by some simply density simulations (on samples), but it's a fairly coarse method, so it wouldn't catch small differences.

Reply:
Based on the derivation in my below comment, it seems that Jacobian has an analytical form:
```python
# |dp/dc| = prod(ds/dx) = prod(expit'(x))
# we know log derivative of expit(x) is `-softplus(x) -softplus(-x)`
J = -(softplus(x) + softplus(-x)).sum(-1)
return J
```
where `x` is the cutpoints. Could you give it a try?

Reply:
It is `cutpoint -> probability`. The Jacobian matrix (which has diagonal + off-diagonal bands) is lower triangular, so its determinant is the product of entries on the diagonal.

> So if anything, maybe we could skip the S=sum(p_i) variable trick (and its column of 1s) and just have a square Jacobian by defining only the p_0, ..., p_{k-1}

Yes, I think this is a right approach. Let's verify it empirically to see if tests pass (i.e. using the Jacobian term in my last comment - note that `x` is the cutpoints in that derivation - and rerun the test).

Reply:
I just did some tests to verify that the above calculation is correct. :)
```python
import numpyro
from numpyro.distributions import constraints, transforms
import jax
import jax.numpy as jnp

# TODO: use this transform in OrderedLogistic
class OrderedToSimplexTransform(transforms.Transform):
    domain = constraints.ordered_vector
    codomain = constraints.simplex

    def __init__(self, anchor_point=0.0):
        self.anchor_point = anchor_point

    def __call__(self, x):
        x = x - jnp.expand_dims(self.anchor_point, -1)
        sigma = jax.nn.sigmoid(x)
        pad_width = [(0, 0)] * (jnp.ndim(sigma) - 1) + [(1, 1)]
        sigma = jnp.pad(sigma, pad_width, constant_values=(0, 1))
        p = sigma[..., 1:] - sigma[..., :-1]
        return jnp.clip(p, a_min=jnp.finfo(p.dtype).tiny)

    def _inverse(self, y):
        sigma = jnp.cumsum(y[..., :-1], axis=-1)
        x = jax.scipy.special.logit(sigma)
        return x + jnp.expand_dims(self.anchor_point, -1)

    def log_abs_det_jacobian(self, x, y, intermediates=None):
        x = x - jnp.expand_dims(self.anchor_point, -1)
        return -(jax.nn.softplus(x) + jax.nn.softplus(-x)).sum(-1)

    def forward_shape(self, shape):
        if len(shape) < 1:
            raise ValueError("Too few dimensions on input")
        return shape[:-1] + (shape[-1] + 1,)

    def inverse_shape(self, shape):
        if len(shape) < 1:
            raise ValueError("Too few dimensions on input")
        return shape[:-1] + (shape[-1] - 1,)


t = transforms.ComposeTransform([transforms.OrderedTransform(),
                                 OrderedToSimplexTransform(),
                                 StickBreakingTransform().inv])
x = jax.random.normal(jax.random.PRNGKey(1), (5,))
expected = jnp.linalg.slogdet(jax.jacobian(t)(x))[1]
actual = t.log_abs_det_jacobian(x, t(x))
print(expected, actual)
# 1.7702796 1.7702808
```

Reply:
As always, you're right - I forgot that we lose the column of 1s, so the zeros under the diagonal will wipe the off-diagonal band.

I've plugged it in as per your suggestion (with `-J_logdet`) given the direction it works in.

Reply:
Sure, you can use the following derivation
```
softplus(x) = -log(sigmoid(-x))
sigmoid'(x) = sigmoid(x)*sigmoid(-x)
=> log (sigmoid'(x)) = -softplus(-x) - softplus(x)
```